compiler/rustc_mir_transform/src/coverage/query.rs

```diff
@@ -38,14 +40,43 @@ fn is_eligible_for_coverage(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         return false;
     }
 
-    if tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::NO_COVERAGE) {
+    if tcx.codegen_fn_attrs(def_id).flags.contains(CodegenFnAttrFlags::NAKED) {
+        trace!("InstrumentCoverage skipped for {def_id:?} (`#[naked]`)");
+        return false;
+    }
+
+    if !tcx.coverage_attr_on(def_id) {
         trace!("InstrumentCoverage skipped for {def_id:?} (`#[coverage(off)]`)");
         return false;
     }
 
     true
 }
 
+/// Query implementation for `coverage_attr_on`.
+fn coverage_attr_on(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
+    // Check for annotations directly on this def.
+    if let Some(attr) = tcx.get_attr(def_id, sym::coverage) {
+        match attr.meta_item_list().as_deref() {
+            Some([item]) if item.has_name(sym::off) => return false,
+            Some([item]) if item.has_name(sym::on) => return true,
+            Some(_) | None => {
+                // Other possibilities should have been rejected by `rustc_parse::validate_attr`.
+                tcx.dcx().span_bug(attr.span, "unexpected value of coverage attribute");
+            }
+        }
+    }
+
+    match tcx.opt_local_parent(def_id) {
+        // Check the parent def (and so on recursively) until we find an
+        // enclosing attribute or reach the crate root.
+        Some(parent) => tcx.coverage_attr_on(parent),
```

Comment:
Is it fine for this query to just call itself recursively (via the query system) to walk up the def tree, or is there something special I have to do here?

Reply:
Presumably it'd make sense to do some sort of path compression/memoisation so that the query is only linear in the number of parents, rather than quadratic. But not quite sure what exists to do that in the compiler.

Reply:
Given that I'm calling `tcx.coverage_attr_on` (and not recursing on `fn coverage_attr_on` directly), my understanding is that the query system will memoize the result.

So functions will only recurse until they hit something (typically the enclosing mod/impl) for which there is a previously-memoized result.

Reply:
Yea this is totally fine. We do this for a lot of queries (e.g. `generics_of`)